source/rc.bind/bind/run

```diff
@@ -0,0 +1,14 @@
+#!/bin/execlineb -P
+
+#############################################
+# Start the BIND name server daemon (named) #
+#############################################
+
+fdmove -c 2 1
+s6-envdir env
+multisubstitute {
+  importas -u CMD CMD
+  importas -usCd" " CMD_OPTS CMD_OPTS
+}
+s6-notifyoncheck -dn0 -c "rndc status"
```

Comment:
Good point. It's been a long time since I've configured bind. If I remember correctly rndc isn't required and named would function correctly without `rndc.key` and `rndc.conf`. `rndc.key`/`rndc.conf` are required to give `rndc` access to `named`, because `rndc` can control `named` over network, they don't always run on the same machine.

> Will there be other service that depends on this service?

Probably not on stock slackware, but I can imagine services that depend on bind. I'm using bind because of rspamd. rspamd does a lot of DNS queries, therefore it requires a caching DNS server like bind. And in my setup bind is a dependency of rspamd. If bind is used for more than caching, then everything that should resolve external domains, would depend on bind.

I'm not sure what is a better way to make a health check. `named` creates a pid file in `/run/named`, but it can be disabled (and I do so, when using s6). `named` creates `/run/named/session.key`, but it can be disabled too, but the file is created with standard configuration. I picked the `rndc` check just because it is the tool for this job. `rndc` is the tool to query and check and control `named`, so it's kind of created for checking `named` status.

Basically I can remove the check. `bind` starts pretty fast and nothing bad happens if it's not yet running, probably some domain names can't be resolved (probably worse for oneshots than longruns). What do you think?

Reply:
> Probably not on stock slackware, but I can imagine services that depend on bind. I'm using bind because of rspamd. rspamd does a lot of DNS queries, therefore it requires a caching DNS server like bind. And in my setup bind is a dependency of rspamd. If bind is used for more than caching, then everything that should resolve external domains, would depend on bind.
> 
> I'm not sure what is a better way to make a health check. `named` creates a pid file in `/run/named`, but it can be disabled (and I do so, when using s6). `named` creates `/run/named/session.key`, but it can be disabled too, but the file is created with standard configuration. I picked the `rndc` check just because it is the tool for this job. `rndc` is the tool to query and check and control `named`, so it's kind of created for checking `named` status.
> 
> Basically I can remove the check. `bind` starts pretty fast and nothing bad happens if it's not yet running, probably some domain names can't be resolved (probably worse for oneshots than longruns). What do you think?

Thanks for your explanations, we'll use `s6-notifyoncheck` then. To avoid `s6-rc -u change` from blocking when the user hasn't configured `rndc` yet, we'll do something like below for the run script:

```
#!/bin/execlineb -P

#############################################
# Start the BIND name server daemon (named) #
#############################################

fdmove -c 2 1
s6-envdir env
multisubstitute {
  importas -u CMD CMD
  importas -usCd" " CMD_OPTS CMD_OPTS
}
ifelse { [ -r /etc/rndc.conf -o -r /etc/rndc.key ] }
{
  s6-notifyoncheck -dn0 -c
  "redirfd -w 2 /dev/null
  pipeline { rndc -s localhost status }
  grep -q \"server is up and running\""
  exec -c $CMD $CMD_OPTS
}
# We can't use rndc to poll the named server readiness so we'll announce
# this service's readiness after some fixed period of time
background -d {
  if { sleep 1 }
  fdmove 1 3
  echo
}
unexport ! fdclose 3
exec -c $CMD $CMD_OPTS
```

So instead of directly go to `s6-notifyoncheck`, we'll test first the existence of `/etc/rndc.conf` or `/etc/rndc.key` and then use `rndc` to poll readiness if one of them exists. Here I change the check script to pipe the output of `rndc` to grep and look for the string "server is up and ready". When `rndc` hasn't configured yet, we'll just start `named` immediately and then wait for 1 second. After that 1 second elapsed, print a newline to the notification fd to announce service readiness.

Will that be good enough for you?

Reply:
Looks good.

Since
```
"redirfd -w 2 /dev/null
  pipeline { rndc -s localhost status }
  grep -q \"server is up and running\""
```
is a long string, can I put it into `data/check`? And if I'm not mistaken s6-rc doesn't mark `data/check` as executable automatically.

Reply:
> Since
> 
> ```
> "redirfd -w 2 /dev/null
>   pipeline { rndc -s localhost status }
>   grep -q \"server is up and running\""
> ```
> 
> is a long string, can I put it into `data/check`?

Yes, feel free to do that.

> And if I'm not mistaken s6-rc doesn't mark `data/check` as executable automatically.

Yes, since s6-rc will only copy the `data` directory verbatim.